processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -109,6 +275,85 @@
     return new GenericIndexed<T>(theBuffer.asReadOnlyBuffer(), strategy, allowReverseLookup);
   }
 
+  public static <T> GenericIndexed<T> read(ByteBuffer buffer, ObjectStrategy<T> strategy)
+  {
+    byte versionFromBuffer = buffer.get();
+
+    if (VERSION_ONE == versionFromBuffer) {
+      return createVersionOneGenericIndexed(buffer, strategy);
+    } else if (VERSION_TWO == versionFromBuffer) {
+      throw new IAE(
+          "use read(ByteBuffer buffer, ObjectStrategy<T> strategy, SmooshedFileMapper fileMapper) to read version 2 indexed.",
+          versionFromBuffer
+      );
+    }
+    throw new IAE("Unknown version[%s]", versionFromBuffer);
+  }
+
+  private static <T> GenericIndexed<T> createVersionOneGenericIndexed(ByteBuffer buffer, ObjectStrategy<T> strategy)
+  {
+    boolean allowReverseLookup = buffer.get() == 0x1;
+    int size = buffer.getInt();
+    ByteBuffer bufferToUse = buffer.asReadOnlyBuffer();
+    bufferToUse.limit(bufferToUse.position() + size);
+    buffer.position(bufferToUse.limit());
+
+    return new GenericIndexed<T>(
+        bufferToUse,
+        strategy,
+        allowReverseLookup
+    );
+  }
+
+  public static <T> GenericIndexed<T> read(ByteBuffer buffer, ObjectStrategy<T> strategy, SmooshedFileMapper fileMapper)
+  {
+    byte versionFromBuffer = buffer.get();
+
+    if (VERSION_ONE == versionFromBuffer) {
+      return createVersionOneGenericIndexed(buffer, strategy);
+    } else {
+      if (VERSION_TWO == versionFromBuffer) {
+
+        if (fileMapper == null) {
+          throw new IAE("SmooshedFileMapper can not be null for version 2.");
+        }
+        boolean allowReverseLookup = buffer.get() == 0x1;
+        int logBaseTwoOfElementsPerValueFile = buffer.getInt();
+        int numWritten = buffer.getInt();
```

Comment:
This is one of the design limitation in GenericIndexed V2 i.e header file can not go beyond 2GB limit. GenericIndexed v2 stores header indexes as int which means we can only store `2^31/2^2=2^29` items which can be stored as int. If header file can have `2^29` items this means GenericIndexed V2 can store up to `2^32 * 2^29` bytes.